src/LtiMessageLaunch.php

```diff
@@ -122,6 +122,49 @@ public static function fromCache(
         return $new->validateRegistration();
     }
 
+    public function initialize(array $request = null): static
+    {
+        $this->setRequest($request);
+
+        // Ideally this would just be abstracted to a validator. But how when, halfway through, we need to migrate?
+        $this->validateState()
+            ->validateJwtFormat()
+            ->validateNonce()
+            ->validateRegistration()
+            ->validateJwtSignature();
+
+        // This feels kinda icky in the middle of all the validation. How can I move it?
+        if ($this->shouldMigrate()) {
+            $this->deployment = $this->db->migrateFromLti11($this->getLaunchData());
+        }
+
+        $this->validateDeployment()
+            ->validateMessage();
+
+        $this->cacheLaunchData();
+
+        return $this;
+    }
+
+    public function setRequest(array $request = null)
+    {
+        if ($request === null) {
+            $request = $_POST;
+        }
+
+        $this->request = $request;
+
+        return $this;
+    }
+
+    private function shouldMigrate(): bool
+    {
+        return $this->hasMigrationStrategy()
+            // I don't like this here. Move it later
+            && isset($this->jwt['body'][LtiConstants::DEPLOYMENT_ID])
+            && $this->db->hasMatchingLti11Key($this->getLaunchData()['oauth_consumer_key_sign']);
```

Comment:
In order to verify this, you need to be passed both `oauth_consumer_key_sign` and `oauth_consumer_key`. These properties will be under the 1.1 migration claim, and not on the top level of the launch data.

Its also possible to not send these values, so you can't assume they are there (referring to the key on the array if its not there will blow up).

Reply:
I threw this together from memory without looking to see where these actually come from or putting in any safeguards. (This definitely isn't meant to be production-ready). Both of your suggestions here sound right to me.